src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,194 @@
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources));
+
+        public AbstractMakeFieldReadonlyDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                   s_localizableTitle, s_localizableMessage)
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        internal void AnalyzeType(SyntaxNodeAnalysisContext context)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferReadonly, context.Node.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var typeSymbol = (ITypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node);
+
+            var nonReadonlyFieldMembers = PooledHashSet<IFieldSymbol>.GetInstance();
+            foreach (var item in typeSymbol.GetMembers())
+            {
+                if (item is IFieldSymbol symbol &&
+                    symbol.DeclaredAccessibility == Accessibility.Private &&
+                    !symbol.IsReadOnly &&
+                    !symbol.IsConst &&
+                    !symbol.IsImplicitlyDeclared &&
+                    !IsMutableValueType(symbol.Type))
+                {
+                    var isInCurrentNode = false;
+                    foreach (var syntaxReference in item.DeclaringSyntaxReferences)
+                    {
+                        if (!syntaxReference.SyntaxTree.Equals(context.Node.SyntaxTree))
+                        {
+                            continue;
+                        }
+
+                        if (!context.Node.Span.Contains(syntaxReference.Span))
+                        {
+                            continue;
+                        }
+
+                        isInCurrentNode = true;
+                        break;
+                    }
+
+                    if (isInCurrentNode)
+                    {
+                        nonReadonlyFieldMembers.Add(symbol);
+                    }
+                }
+            }
+
+            var syntaxFactsService = GetSyntaxFactsService();
+            foreach (var syntaxReference in typeSymbol.DeclaringSyntaxReferences)
+            {
+                var typeNode = syntaxReference.SyntaxTree.GetRoot(context.CancellationToken).FindNode(syntaxReference.Span);
```

Comment:
isn't this just syntaxReference.GetSyntax(cancellationToken)?

Reply:
:memo: I tried to make this change, but it broke a bunch of the VB tests. I have no idea why.